### PR TITLE
Prepare certbot module for mypy check untyped defs

### LINKED ISSUE #5952: Check untyped defs in the certbot module
This issue is for checking untyped definitions in the `certbot` module.

See explanation here: https://docs.google.com/document/d/1fp0tPP9bzYkUViTnOXAyoT_KfgTfs8HSpNMoU8Tm_SY/edit?usp=sharing

From http://blog.zulip.org/2016/10/13/static-types-in-python-oh-mypy/:

Add the `--check-untyped-defs` option to your mypy arguments, and get that running with no errors on the codebase. This option causes mypy to check every def in the codebase for internal consistency; mypy can detect many classes of bugs and mistakes in your codebase, without your having written a single type annotation!

In most cases, you’ll want to fix bugs and bad code as you go, but you can also use #type: ignore annotations or exclude files to defer issues. For example, we excluded all of Zulip’s tests at first, since they’re both lower value to type-check and had most of the monkey-patching and other questionable Python in the project. For Zulip, getting clean `--check-untyped-defs` output took me about 2 days of hard work, including merging fixes for about 40 issues in the Zulip codebase.

I spent another day or two generating nice reproducers for the mypy bugs I’d encountered and improving typeshed. Now that mypy is no longer in its infancy, mypy bugs are increasingly rare. But a large project should expect to encounter and fix typeshed issues (just submit a PR!).

Once `mypy --check-untyped-defs` passes on your codebase, you should lock down your progress by running the mypy checker in your CI environment.

Because mypy type annotations are optional, once you’ve done the setup above, you can start annotating your codebase at whatever pace you like. Over time, you’ll get the benefits of static types in the parts of the codebase that you’ve annotated, without needing to do anything special to the rest of the codebase (the wonders of gradual typing!). In the next section, I’ll discuss strategy for how to move the codebase towards being fully annotated.

### PR DESCRIPTION
**Pull request by dmfigol**
Fix #5952

### CODE REVIEW

**Comment by dmfigol** [comment]
It is all green!
CC @bmw

File: certbot/plugins/standalone.py [hunk 1]
```diff
@@ -3,21 +3,30 @@
...
 import socket
+from socket import errno as socket_errors  # type: ignore
 
 import OpenSSL
 import six
 import zope.interface
 
 from acme import challenges
 from acme import standalone as acme_standalone
+# pylint: disable=unused-import, no-name-in-module
```
Comment (bmw) [inline certbot/plugins/standalone.py]: Same comment about inline `pylint: disable`.

File: certbot/plugins/standalone.py [hunk 2]
```diff
@@ -179,14 +190,14 @@ def __init__(self, *args, **kwargs):
...
-        self.served = collections.defaultdict(set)
+        self.served = collections.defaultdict(set)  # type: ServedType
 
         # Stuff below is shared across threads (i.e. servers read
         # values, main thread writes). Due to the nature of CPython's
         # GIL, the operations are safe, c.f.
         # https://docs.python.org/2/faq/library.html#what-kinds-of-global-value-mutation-are-thread-safe
-        self.certs = {}
-        self.http_01_resources = set()
+        self.certs = {}  # type: Dict[bytes, Tuple[OpenSSL.crypto.PKey, OpenSSL.crypto.PKey]]
```
Comment (bmw) [inline certbot/plugins/standalone.py]: I think the last type in the tuple is `OpenSSL.crypto.X509`.

File: certbot/plugins/standalone.py [hunk 3]
```diff
@@ -179,14 +190,14 @@ def __init__(self, *args, **kwargs):
...
+        self.http_01_resources = set()  # type: Set[challenges.HTTP01Response]
```
Comment (bmw) [inline certbot/plugins/standalone.py]: Isn't this `Set[acme_standalone.HTTP01RequestHandler.HTTP01Resource]`?

File: certbot/plugins/standalone.py [hunk 4]
```diff
@@ -3,21 +3,30 @@
 import collections
 import logging
 import socket
+from socket import errno as socket_errors  # type: ignore
```
Comment (bmw) [inline certbot/plugins/standalone.py]: Is there an issue tracking the need to use `type: ignore` here? Can we link it?

File: certbot/plugins/standalone_test.py [hunk 5]
```diff
@@ -2,12 +2,16 @@
 import argparse
 import socket
 import unittest
+from socket import errno as socket_errors  # type: ignore
```
Comment (bmw) [inline certbot/plugins/standalone_test.py]: Is there an issue tracking the need to use `type: ignore` here? Can we link it?

File: certbot/plugins/standalone_test.py [hunk 6]
```diff
@@ -21,8 +25,8 @@ class ServerManagerTest(unittest.TestCase):
 
     def setUp(self):
         from certbot.plugins.standalone import ServerManager
-        self.certs = {}
-        self.http_01_resources = {}
+        self.certs = {}  # type: Dict[bytes, Tuple[OpenSSL.crypto.PKey, OpenSSL.crypto.PKey]]
```
Comment (bmw) [inline certbot/plugins/standalone_test.py]: I think the last type in the tuple is `OpenSSL.crypto.X509`.

File: certbot/plugins/standalone_test.py [hunk 7]
```diff
@@ -21,8 +25,8 @@ class ServerManagerTest(unittest.TestCase):
...
+        self.http_01_resources = {}  # type: Set[challenges.HTTP01Response]
```
Comment (bmw) [inline certbot/plugins/standalone_test.py]: Isn't this `Set[acme_standalone.HTTP01RequestHandler.HTTP01Resource]`?

File: certbot/plugins/webroot.py [hunk 8]
```diff
@@ -10,8 +10,11 @@
 import zope.component
 import zope.interface
 
-from acme import challenges
+from acme import challenges  # pylint: disable=unused-import,
+# pylint: disable=unused-import, no-name-in-module
```
Comment (bmw) [inline certbot/plugins/webroot.py]: Same comment about preferring inline `pylint` comments.

File: certbot/tests/error_handler_test.py [hunk 9]
```diff
@@ -6,6 +6,8 @@
 import unittest
 
 import mock
+# pylint: disable=unused-import, no-name-in-module
```
Comment (bmw) [inline certbot/tests/error_handler_test.py]: Same comment about preferring inline `pylint`.

File: certbot/tests/error_handler_test.py [hunk 10]
```diff
@@ -23,7 +25,7 @@ def set_signals(sig_handler_dict):
 def signal_receiver(signums):
     """Context manager to catch signals"""
     signals = []
-    prev_handlers = {}
+    prev_handlers = {}  # type: Dict[int, Union[int, None, Callable[[], None]]]
```
Comment (bmw) [inline certbot/tests/error_handler_test.py]: See my comment on `error_handler.py` about what I think the type of the callable here should be.

File: acme/acme/magic_typing.py [hunk 11]
```diff
@@ -9,5 +9,6 @@ def __getattr__(self, name):
 try:
     # mypy doesn't respect modifying sys.modules
     from typing import * # pylint: disable=wildcard-import, unused-wildcard-import
+    from typing import IO  # pylint: disable=unused-import
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply acme/acme/magic_typing.py]: done

File: certbot/auth_handler.py [hunk 12]
```diff
@@ -117,7 +117,7 @@ def _has_challenges(self, aauthzrs):
 
     def _solve_challenges(self, aauthzrs):
         """Get Responses for challenges from authenticators."""
-        resp = []
+        resp = []  # type: List[str]
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply certbot/auth_handler.py]: done

File: certbot/client.py [hunk 13]
```diff
@@ -5,7 +5,7 @@
 import platform
 
 from cryptography.hazmat.backends import default_backend
-from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.hazmat.primitives.asymmetric.rsa import generate_private_key  # type: ignore
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply certbot/client.py]: not added to typeshed

File: certbot/client.py [hunk 14]
```diff
@@ -605,8 +605,10 @@ def validate_key_csr(privkey, csr=None):
         if csr.form == "der":
             csr_obj = OpenSSL.crypto.load_certificate_request(
                 OpenSSL.crypto.FILETYPE_ASN1, csr.data)
-            csr = util.CSR(csr.file, OpenSSL.crypto.dump_certificate(
-                OpenSSL.crypto.FILETYPE_PEM, csr_obj), "pem")
+            cert_buffer = OpenSSL.crypto.dump_certificate(  # type: ignore
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply certbot/client.py]: done

File: certbot/crypto_util.py [hunk 15]
```diff
@@ -8,15 +8,17 @@
...
-import OpenSSL
+
 import pyrfc3339
 import six
 import zope.component
+from OpenSSL import crypto
+from OpenSSL import SSL  # type: ignore
 from cryptography.hazmat.backends import default_backend
-from cryptography import x509 # type: ignore
+from cryptography import x509  # type: ignore
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply certbot/crypto_util.py]: https://github.com/python/typeshed/tree/master/third_party/2/cryptography x509 is not in typeshed

File: certbot/error_handler.py [hunk 16]
```diff
@@ -56,9 +59,9 @@ class ErrorHandler(object):
     def __init__(self, func=None, *args, **kwargs):
         self.call_on_regular_exit = False
         self.body_executed = False
-        self.funcs = []
-        self.prev_handlers = {}
-        self.received_signals = []
+        self.funcs = []  # type: List[Callable[[], None]]
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply certbot/error_handler.py]: done

File: certbot/error_handler.py [hunk 17]
```diff
@@ -56,9 +59,9 @@ class ErrorHandler(object):
...
+        self.prev_handlers = {}  # type: Dict[int, Union[int, None, Callable[[], None]]]
```
[the remark this answers is not in the document]
Comment (dmfigol) [reply certbot/error_handler.py]:
```
In [1]: import signal

In [2]: signal.getsignal??
Docstring:
getsignal(sig) -> action

Return the current action for the given signal.  The return value can be:
SIG_IGN -- if the signal is being ignored
SIG_DFL -- if the default action for the signal is in effect
None -- if an unknown handler is in effect
anything else -- the callable Python object used as a handler
Type:      builtin_function_or_method
```

[the remark this answers is not in the document]
Comment (dmfigol) [reply certbot/error_handler.py]: will change to
`self.prev_handlers = {}  # type: Dict[int, Union[int, None, Callable]]`

File: certbot/plugins/standalone.py (hunk 4 above)
Comment (dmfigol) [reply certbot/plugins/standalone.py]: https://github.com/python/typeshed/blob/master/stdlib/2and3/socket.pyi

File: certbot/plugins/standalone.py (hunk 1 above)
Comment (dmfigol) [reply certbot/plugins/standalone.py]: done

File: certbot/plugins/standalone.py (hunk 2 above)
Comment (dmfigol) [reply certbot/plugins/standalone.py]: yes, fixed

File: certbot/plugins/standalone.py (hunk 3 above)
Comment (dmfigol) [reply certbot/plugins/standalone.py]: `acme_standalone.HTTP01Resource` fixed

File: certbot/plugins/standalone_test.py (hunk 5 above)
Comment (dmfigol) [reply certbot/plugins/standalone_test.py]: # https://github.com/python/typeshed/blob/master/stdlib/2and3/socket.pyi

Comment (dmfigol) [reply certbot/plugins/standalone_test.py]: https://github.com/python/typeshed/blob/master/stdlib/2and3/socket.pyi

File: certbot/plugins/standalone_test.py (hunk 6 above)
Comment (dmfigol) [reply certbot/plugins/standalone_test.py]: yes, fixed

File: certbot/plugins/standalone_test.py (hunk 7 above)
Comment (dmfigol) [reply certbot/plugins/standalone_test.py]: fixed

File: certbot/tests/error_handler_test.py (hunk 10 above)
Comment (dmfigol) [reply certbot/tests/error_handler_test.py]: yes, fixed

File: mypy.ini [hunk 18]
```diff
@@ -31,6 +37,3 @@ check_untyped_defs = True
 
 [mypy-certbot_dns_digitalocean.*]
 check_untyped_defs = True
-
```
Comment (bmw) [inline mypy.ini]: Looks like something happened with the merge here. These lines for nginx shouldn't be deleted.

**Pull request merged**